### PR TITLE
Ensure action plugins accept only valid args

### LINKED ISSUE #44773: Pure action plugins accept every parameter without error
##### SUMMARY
As explained in #25424, almost all pure action plugins take whatever parameter it needs, but doesn't report back unsupported parameters. As a result, mis-indented task-directives end up being ignored.

Affected plugins are:
- add_host *-- Plugin accepts any parameter to become a host variable*
- assert
- debug *-- Already using VALID_ARGS*
- fail
- group_by
- include_vars *-- Implemented own comprehensive fix*
- pause *-- Implemented own fix*
- reboot
- set_fact *-- Plugin accepts any parameter to become a "fact"*
- set_stats
- wait_for_connection
- win_reboot

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
assert, fail, group_by, pause, reboot, set_stats, wait_for_connection, win_reboot

##### ANSIBLE VERSION
v2.7 and older

##### STEPS TO REPRODUCE
Very simple example
```yaml
  - fail:
      msg: 'Something failed'
      when: 1 == 2
```

##### EXPECTED RESULTS
A proper message that 'when' is not a valid parameter.

##### ACTUAL RESULTS
The task works where the user would have expected it not to work.
Total chaos and panic ! Ansible is being ditched for Puppet or Chef.

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
This fixes #25424
This also fixes #44773

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
assert, fail, group_by, pause, reboot, set_stats, wait_for_connection, win_reboot

##### ANSIBLE VERSION
v2.7 and older

### CODE REVIEW

**Review by bcoca — APPROVED** [review]

**Review by acozine — APPROVED** [review]
Thanks for fixing typos while you were changing the code!

**Review by bcoca — APPROVED** [review]

File: lib/ansible/plugins/action/win_reboot.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ class TimedOutException(Exception):
 
 class ActionModule(RebootActionModule, ActionBase):
     TRANSFERS_FILES = False
+    VALID_ARGS = frozenset(('connect_timeout', 'msg', 'post_reboot_delay', 'post_reboot_delay_sec', 'pre_reboot_delay', 'pre_reboot_delay_sec', 'test_command'))
```
Comment (samdoran) [inline lib/ansible/plugins/action/win_reboot.py]: This should also accept the deprecated args.

**Comment by abadger** [comment]
Change it to do it inside of each module instead of a s a generic service and we can merge this.  sdoran is going to be working on controller-side arg_spec in the next release and we don't want to keep adding things to the action plugin API that we then have to support for a long time despite not wanting to do it that way anymore.

**Comment by samdoran** [comment]
Please create a changelog fragment. See [fragments](https://github.com/ansible/ansible/tree/stable-2.6/changelogs/fragments) for examples.

**Comment by abadger** [comment]
@dagwieers Actually... slightly less work... Mark the variable as private: ```_VALID_ARGS```.  Then we can feel free to change this when we implement controller-side argspec and any 3rd party modules shouldn't have been relying on it in the first place.

File: lib/ansible/plugins/action/win_reboot.py (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/plugins/action/win_reboot.py]: It seems I missed some...
But I'd rather deprecate them for real so we can get rid of them in 4 releases.

**Comment by dagwieers** [comment]
Done. Please re-review.

**Comment by abadger** [comment]
Looks good to me.

**Pull request merged**